code/engine/ecs/archetype.lua

```diff
@@ -0,0 +1,138 @@
+local CACHE_WITH = {}
+local CACHE_WITHOUT = {}
+local archetypes = {} -- All the archetypes
+local version = 0 -- Archetype version, changes whenever a new type of archetype is added
+
+local archetype = {}
+archetype.__index = archetype
+
+-- Gets already existing archetype or creates a new one
+function archetype.setup(...)
+  local component_types = (...).is_component_type and { ... } or ...
+  local component_ids = {}
+  local components = {}
+
+  for _, component in pairs(component_types) do
+    if (component.is_component_type and not component.is_component) then
+      if components[component] == nil then
+        components[component] = true
+        table.insert(component_ids, component:get_id())
+      end
+    end
+  end
+
+  local archetype_id = table.concat(component_ids, "_")
+
+  if archetypes[archetype_id] == nil then
+    archetypes[archetype_id] = setmetatable({
+      _id = archetype_id,
+      _components = components,
+    }, archetype)
+
+    version = version + 1
+  end
+
+  return archetypes[archetype_id]
+end
+
+-- Adds a component to a archetype
+function archetype:add(component_type)
+  if self._components[component_type] then
+    return self
+  end
+
+  local archetype_cache = CACHE_WITH[self]
+
+  if not archetype_cache then
+    archetype_cache = {}
+    CACHE_WITH[self] = archetype_cache
+  end
+
+  local current_archetype = archetype_cache[component_type]
+
+  if current_archetype == nil then
```

Comment:
nil checks på två olika sätt (rad 44 var det "if not" å här "== nil" , kanske borde va på samma sätt?